### DISCUSSION TITLE
XMLSec 1.3.0-rc3

### DISCUSSION THREAD

**Discussion by lsh123** [lsh123/xmlsec#612] [Announcements]
Download: [source](https://www.aleksey.com/xmlsec/download/xmlsec1-1.3.0-rc3.tar.gz), [sig](https://www.aleksey.com/xmlsec/download/xmlsec1-1.3.0-rc3.sig)

## Changes from 1.3.0-RC2
-   xmlsec-core: Added failure reason to xmlSecDSigCtx and xmlSecEncCtx;
-   all xmlsec-crypto libraries: Ensured that only a single copy of each certificate is added to the key;
-   xmlsec-nss: Improved certificates search and allowed self-signed certificates (if certificate verification is not required).

## Changes from 1.3.0-RC1
* Added missed `include/xmlsec/x509.h` (thanks @vmiklos for bug report)
* Updated README file to new version LibGnuTLS

## XMLSec 1.3.0-RC1
The XMLSec 1.3.0 release includes a large number of changes including several API / ABI breaking changes (hence version bump). Please test the [release candidate](https://www.aleksey.com/xmlsec/download/xmlsec1-1.3.0-rc1.tar.gz) and let me know if you run into any issues! I plan to have at least one more release candidate in 2-3 weeks before the release mid-April.

### core xmlsec and all xmlsec-crypto libraries:
* **(ABI breaking change)** Added support for the [KeyInfoReference Element](https://www.w3.org/TR/xmldsig-core1/#sec-KeyInfoReference).
* **(ABI breaking change)** Switched xmlSecSize to use size_t by default. Use "--enable-size-t=no" configure option ("size_t=no" on Windows) to restore the old behaviour (note that support for xmlSecSize being different from size_t will be removed in the future).
* **(API breaking change)** Changed the key search to strict mode: only keys referenced by KeyInfo are used. To restore the old "lax" mode, set XMLSEC_KEYINFO_FLAGS_LAX_KEY_SEARCH flag on xmlSecKeyInfoCtx or use '--lax-key-search' option for XMLSec command line utility.
* **(API breaking change)** The KeyName element content is now trimmed before key search is performed.
* **(API breaking change)** Disabled FTP support by default. Use "--enable-ftp" configure option to restore it. Also added "--enable-http" and "--enable-files" configure options to control support for loading files over HTTP or locally.
* **(API/ABI breaking change)** Disabled MD5 digest method by default. Use "--enable-md5" configure options ("legacy-crypto" option on Windows) to re-enable MD5.
* **(ABI breaking change)** Removed deprecated functions.
* Added support for loading keys through [ossl-store](https://www.openssl.org/docs/man3.0/man7/ossl_store.html) interface (e.g. for using keys from an HSM). Also see '--privkey-openssl-store' and '--pubkey-openssl-store ' command line options for XMLSec utility.
* Added ability to control transforms binary chunk size to improve performance (see '--transform-binary-chunk-size' command line option for XMLSec utility).
* Fixed all potentially unsafe integer conversions and all the other warnings.
* Added [XML Signature 1.1 interop (2012)](https://www.w3.org/TR/2012/NOTE-xmldsig-core1-interop-20121113/) and [XML Encryption 1.1 interop (2012)](https://www.w3.org/TR/2012/NOTE-xmlenc-core1-interop-20121113/) tests.

### xmlsec-openssl library:
* Added support for [SHA3 digests](https://www.ietf.org/rfc/rfc9231.html#name-sha-3-algorithms).
* Added support for [ECDSA-SHA3 signatures](https://www.ietf.org/rfc/rfc9231.html#name-ecdsa-sha-ecdsa-ripemd160-e).
* Added support for [RSA PSS signatures (withtout parameters)](https://www.ietf.org/rfc/rfc9231.html#section-2.3.10).
* Added support for [ConcatKDF key](https://www.w3.org/TR/xmlenc-core1/#sec-ConcatKDF) and [PBKDF2](https://www.w3.org/TR/xmlenc-core1/#sec-PBKDF2) derivation algorithms.
* **(ABI breaking change)** Added support for [ECDH-ES Key Agreement algorithm](https://www.w3.org/TR/xmlenc-core1/#sec-ECDH-ES).
* **(ABI breaking change)** Added support for [DH-ES Key Agreement algorithm](https://www.w3.org/TR/xmlenc-core1/#sec-DHKeyAgreementExplicitKDF) with explicit KDF.
* Added support for [MGF1 algorithm to RSA OAEP key transport](https://www.w3.org/TR/xmlenc-core1/#sec-RSA-OAEP).
* Added support for [X509Digest](https://www.w3.org/TR/xmldsig-core1/#sec-X509Data) element and ability to lookup keys using other X509Data elements.
* Added support for [DEREncodedKeyValue](https://www.w3.org/TR/xmldsig-core1/#sec-DEREncodedKeyValue) element.
* Automatically set key name from PKCS12 key name.
* Removed support for OpenSSL 1.0.0 and LibreSSL before 2.7.0.

### xmlsec-nss library:
* Added support for [RSA PSS signatures (withtout parameters)](https://www.ietf.org/rfc/rfc9231.html#section-2.3.10).
* Added support for [RSA OAEP key transport including MGF1 algorithms](https://www.w3.org/TR/xmlenc-core1/#sec-RSA-OAEP).
* Added support for [AES GCM ciphers](https://www.w3.org/TR/xmlenc-core1/#sec-AES-GCM).
* Added support for [PBKDF2](https://www.w3.org/TR/xmlenc-core1/#sec-PBKDF2) derivation algorithms.
* Added support for [X509Digest](https://www.w3.org/TR/xmldsig-core1/#sec-X509Data) element and ability to lookup keys using other X509Data elements.
* Added support for [DEREncodedKeyValue](https://www.w3.org/TR/xmldsig-core1/#sec-DEREncodedKeyValue) element.
* Automatically set key name from PKCS12 key name.

### xmlsec-gnutls library:
* **(API/ABI breaking change)** Removed dependency on xmlsec-gcrypt and libgcrypt libraries (including API functions) to enable support for different GnuTLS backends.
* Bumped minimal GnuTLS version to 3.6.13.
* Added support for [SHA3 digests](https://www.ietf.org/rfc/rfc9231.html#name-sha-3-algorithms).
* Added support for [ECDSA signatures](https://www.w3.org/TR/xmldsig-core1/#sec-ECDSA).
* Added support for [DSA-SHA256 signatures](https://www.w3.org/TR/xmlenc-core1/#sec-RSA-OAEP).
* Added support for [RSA PSS signatures (withtout parameters)](https://www.ietf.org/rfc/rfc9231.html#section-2.3.10).
* Added support for [RSA PKCS 1.5 key transport](https://www.w3.org/TR/xmlenc-core1/#sec-RSA-1_5).
* Added support for [AES GCM ciphers](https://www.w3.org/TR/xmlenc-core1/#sec-AES-GCM).
* Added support for [PBKDF2](https://www.w3.org/TR/xmlenc-core1/#sec-PBKDF2) derivation algorithms.
* Added support for [X509Digest](https://www.w3.org/TR/xmldsig-core1/#sec-X509Data) element and ability to lookup keys using other X509Data elements.
* Added support for [DEREncodedKeyValue](https://www.w3.org/TR/xmldsig-core1/#sec-DEREncodedKeyValue) element.
* Automatically set key name from PKCS12 key name.

### xmlsec-mscng library:
* Added support for [RSA PSS signatures (withtout parameters)](https://www.ietf.org/rfc/rfc9231.html#section-2.3.10).
* Added support for [MGF1 algorithm to RSA OAEP key transport](https://www.w3.org/TR/xmlenc-core1/#sec-RSA-OAEP).
* **(ABI breaking change)** Added support for [ECDH-ES Key Agreement algorithm](https://www.w3.org/TR/xmlenc-core1/#sec-ECDH-ES).
* Added support for [ConcatKDF key](https://www.w3.org/TR/xmlenc-core1/#sec-ConcatKDF) and [PBKDF2](https://www.w3.org/TR/xmlenc-core1/#sec-PBKDF2) derivation algorithms.
* Added support for [X509Digest](https://www.w3.org/TR/xmldsig-core1/#sec-X509Data) element for keys and certificates lookup from the system stores (only SHA1 is supported).
* Added support for [DEREncodedKeyValue](https://www.w3.org/TR/xmldsig-core1/#sec-DEREncodedKeyValue) element.
* Automatically set key name from PKCS12 key name.

### xmlsec-mscrypto library:
* In maintenance mode starting from this release.
* Disabled by default support for NT4. Use "nt4=yes" configure option on Windows to re-enable it.

### xmlsec-gcrypt library:
* In maintenance mode starting from this release.
* Added support for [SHA3 digests](https://www.ietf.org/rfc/rfc9231.html#name-sha-3-algorithms).
* Added support for [ECDSA signatures](https://www.w3.org/TR/xmldsig-core1/#sec-ECDSA).
* Added support for [RSA PSS signatures (withtout parameters)](https://www.ietf.org/rfc/rfc9231.html#section-2.3.10).
* Added support for [RSA PKCS 1.5 key transport](https://www.w3.org/TR/xmlenc-core1/#sec-RSA-1_5).
* Added support for [RSA OAEP key transport including MGF1 algorithms](https://www.w3.org/TR/xmlenc-core1/#sec-RSA-OAEP).

### xmlsec command line utility:
* **(API breaking change)** The XMLSec command line utility is using 'strict' key search mode by default. To restore the old 'lax' key search mode, use the new '--lax-key-search' option.
* Added '--transform-binary-chunk-size' option to control transforms binary chunk size (increasing the chunk size should improve performance at the expense of memory usage.
* Added support for loading keys through [ossl-store](https://www.openssl.org/docs/man3.0/man7/ossl_store.html) interface (e.g. for using keys from an HSM). Also see '--privkey-openssl-store' and '--pubkey-openssl-store ' command line options for XMLSec utility.
* Added '--enabled-key-info-reference-uris' option to control processing of the the [KeyInfoReference Element](https://www.w3.org/TR/xmldsig-core1/#sec-KeyInfoReference).
* Added '--pbkdf2-key' option for loading PBKDF2 keys.
* Added '--concatkdf-key' option for loading ConcatKDF keys.
* Added '--hmac-min-out-len' option to control the min accepted HMAC Output length.
* Added '--pubkey-openssl-engine' option to load public keys from OpenSSL engine.
* Added '--crl-pem' and '--crl-der' options to load CRLs.
* Added '--verify-keys' option to verify key's certificate before loading into Keys Manager (only supported for OpenSSL currently).
* Enabled templatized output filenames to facilitate batch operations on multiple input files.

<hr /><em>This discussion was created from the release <a href='https://github.com/lsh123/xmlsec/releases/tag/xmlsec_1_3_0_rc3'>1.3.0-rc3</a>.</em>

**Comment by vmiklos** [comment]
FWIW I ran the libreoffice integration tests on Windows with the mscng backend and found no problems, sounds like we already sorted out all problems I can see with the nss testing of rc2. :-)

**Reply by lsh123** [reply]
Thanks for checking!

**Comment by postboy** [comment]
"xmlsec-mscrypto library:
    In maintenance mode starting from this release."

Could you please elaborate on that? Does it means that patches and/or bugs regarding this library is no longer considered? Or maybe that means that old code is still actively supported but no new functionality is planned for adding?

**Reply by lsh123** [reply]
Maintenance mode means that there will be no new features, and only critical bug fixes. In the future, xmlsec-mscrypto might be completely deprecated and removed though but I have no specific plans or timeline for this. 

Any specific reasons why xmlsec-mscng doesn’t work for your use case? It should be supported by all Windows releases in the last 20 years ;)

**Reply by postboy** [reply]
Thank you very much for clarification.

Windows CryptoAPI is sometimes reimplemented by companies other than Microsoft for platforms other than Windows. I work for one of those companies. xmlsec-mscrypto can be compiled to be used on *nix systems together with our CryptoAPI implementation. At this moment we don't have CNG API implementation on *nix.

**Reply by lsh123** [reply]
I see. Well, as I mentioned, I don’t have any specific plans to remove the code but in general MSCrypto API is old and doesn’t support newer crypto algorithms. As such it will become obsolete sooner than later.
